runtime-transaction/src/transaction_meta.rs

```diff
@@ -11,13 +11,29 @@
 //! The StaticMeta and DynamicMeta traits are accessor traits on the
 //! RuntimeTransaction types, not the TransactionMeta itself.
 //!
-use solana_sdk::hash::Hash;
+use {
+    solana_program_runtime::compute_budget_processor::ComputeBudgetLimits,
+    solana_sdk::{hash::Hash, slot_history::Slot},
+};
+
+/// private trait to avoid exposing `ComputeBudgetLimits` to callsite of RuntimeTransaction.
+pub(crate) trait RequestedLimits {
+    fn requested_limits(&self, current_slot: Option<Slot>) -> Option<&ComputeBudgetLimits>;
+}
 
 /// metadata can be extracted statically from sanitized transaction,
-/// for example: message hash, simple-vote-tx flag, compute budget limits,
-pub trait StaticMeta {
+/// for example: message hash, simple-vote-tx flag, limits set by instructions
+#[allow(private_bounds)]
+pub trait StaticMeta: RequestedLimits {
     fn message_hash(&self) -> &Hash;
     fn is_simple_vote_tx(&self) -> bool;
+
+    // get fields' value from RequestedLimitsWithExpiry,
+    // `current_slot`, sets to None to skip expiry check, otherwise if current_slot is
+    // before expiry, Some(value) is returned, otherwise return None.
+    fn compute_unit_limit(&self, current_slot: Option<Slot>) -> Option<u32>;
+    fn compute_unit_price(&self, current_slot: Option<Slot>) -> Option<u64>;
+    fn loaded_accounts_bytes(&self, current_slot: Option<Slot>) -> Option<u32>;
```

Comment:
Just gonna throw some initial thoughts down.

1. Naming seems off, this new meta data isn't really static if it expires.
2. Almost seems like we really have 3 types of meta (static, feature-dependent, dynamic)?
3. Having checked-access on each field seems really inconvenient to use

If we introduce some sort of new meta that does expire. It'd be ideal to just have some wrapped reference returned w/ the trait impl'd.

Some dummy code below with poor naming, but I tihnk this might be a more convenient interface.
Basically, you check the expiration on the entire meta and get back a type-safe wrapped reference to the meta which gives access to the expiring meta fields.

This also make the interface for the unchecked path more obvious, instead of passing `None` imo.
As you could just have some separate fn that's `fn unchecked_expiring_meta() -> ...`

For example if I want to get both cu price and cu limit:
```rust
    let Some(expiring_meta) = tx.expire_checked_meta() else {
       // Do something if expired
    }
    let priority_fee = expiring_meta.cu_price + expiring_meta.cu_limit;
```

## Dummy Code

```rust
pub struct Meta {
    a: u64,
    expiration: u64,
}

pub trait ExpiringMeta {
    fn a(&self) -> u64;
}


pub struct ExpireCheckedMeta<'a>(&'a Meta);
impl<'a> ExpiringMeta for ExpireCheckedMeta<'a> {
    fn a(&self) -> u64 {
        self.0.a
    }
}

pub trait ExpiringMetaAccessor {
    fn expire_checked_meta<'a>(&'a self, current: u64) -> Option<ExpireCheckedMeta<'a>>;
}

pub struct RuntimeTransaction<M> {
    m: M,
    meta: Meta,
}

impl<M> ExpiringMetaAccessor for RuntimeTransaction<M> {
    fn expire_checked_meta<'a>(&'a self, current: u64) -> Option<ExpireCheckedMeta<'a>> {
        if self.meta.expiration > current {
            Some(ExpireCheckedMeta(&self.meta))
        } else {
            None
        }
    }
}
```

Reply:
just a quick thought on:

1. agree that passing `option<slot>` as parameter to imply if want to check expiry or not if a hack. like just use `checked_/unchecked_` naming.

2. wrt
> For example if I want to get both cu price and cu limit
even  cu price and cu limit can be used together, but they don't have to be related at all. They in fact are two entirely separate terms. I try not to bundle them together into one entity.

But thanks for the suggestions, it's great to give this one a good look. Cause the next step is to actually use `RuntimeTransaction`, then changes will be more involved.

Reply:
Yeah, I didn't mean to imply that those were necessarily related. But all these "expire checked" fields would be expiring on the same slot, right? I meant that it'd be nice to just check that once if I want to access multiple fields, instead of needing to do it for each one.

```rust
   let Some(checked_meta) = tx.expire_checked_meta(slot) else {/*...*/};
   let a = checked_meta.a();
   let b = checked_meta.b();
   // do stuff
```

vs

```rust
   let Some(a) = tx.checked_a(slot) else { /*...*/ };
   let Some(b) = tx.checked_b(slot) else { /*...*/ };
```

Assuming `/*...*/` is not really simple, this gets repetitive, and does the same check multiple times.

Reply:
still feel unsatisfying to bundle them just because they share same expiry. How about just check expiry at tx itself?

```
let meta_expired: bool = tx.check_expiry(slot);
if do_not_care_expiry || !meta_expired {
    let a = tx.a();
} else {
/*...renew and ...*/
}
```

Reply:
In that, there's no type-safety. I don't need to `check_expiry()` before calling `a()`.

Reply:
i didn't like that looseness either. Just want to throw it out anyway ;) 

Looks like this whole TTL thing is a thing no more - once https://github.com/solana-labs/solana/pull/34472 lands.

If we still want to be support this feature for future expiry-able data, I'm ready tyo take bundled approach at https://github.com/solana-labs/solana/pull/34361/files#r1425677816

Reply:
> Looks like this whole TTL thing is a thing no more - once https://github.com/solana-labs/solana/pull/34472 lands.

Sweet, yeah if this stuff is **actually** static now once 34472 lands, then let's just merge this without expiration after that lands.

Reply:
rebased 34472, now all requested limits are flat fields of static meta.